README
NAME
    DB::Pluggable - Add plugin support for the Perl debugger

VERSION
    version 1.101051

SYNOPSIS
        $ cat ~/.perldb

        use DB::Pluggable;
        use Hook::Modular::Builder;
        my $config = builder {
            log_level 'error';
            enable 'BreakOnTestNumber';
            enable 'StackTraceAsHTML';
            enable 'TypeAhead', type => [ '{l', 'c' ] if $ENV{DBTYPEAHEAD};
            enable 'Dumper';
        };

        $DB::PluginHandler = DB::Pluggable->new(config => $config);
        $DB::PluginHandler->run;

    Alternatively, build the configuration yourself, for example, using
    YAML:

        $ cat ~/.perldb

        use DB::Pluggable;
        use YAML;

        $DB::PluginHandler = DB::Pluggable->new(config => Load <<EOYAML);
        global:
          log:
            level: error

        plugins:
          - module: BreakOnTestNumber
        EOYAML

        $DB::PluginHandler->run;

    Then:

        $ perl -d foo.pl

DESCRIPTION
    This class adds plugin support to the Perl debugger. It is based on
    Hook::Modular, so see its documentation for details.

    You need to have a "~/.perldb" file (see perldebug for details) that
    invokes the plugin mechanism. The one in the synopsis will do, and there
    is a more commented one in this distribution's "etc/perldb" file.

    Plugins should live in the "DB::Pluggable::" namespace, like
    DB::Pluggable::BreakOnTestNumber does.

METHODS
  enable_watchfunction
    Tells the debugger to call "DB::watchfunction()", which in turn calls
    the "db.watchfunction" hook on all plugins that have registered it.

  run
    First it calls the "plugin.init" hook, then it enables hooks for the
    relevant debugger commands (see above for which hooks are available).

    Each command-related hook should return the appropriate constant from
    DB::Pluggable::Constants - either "HANDLED" if the hook has handled the
    command, or "DECLINED" if it didn't. If no hook has "HANDLED" the
    command, the default command subroutine (e.g., "DB::cmd_b()") from
    "perl5db.pl" will be called.

HOOKS
    This class is very much in beta, so it's more like a proof of concept.
    Therefore, not all hooks imaginable have been added, only the ones to
    make this demo work. If you want more hooks or if the current hooks
    don't work for you, let me know.

    The following hooks exist:

    "plugin.init"
        Called at the beginning of the "run()" method. The hook doesn't get
        any arguments.

    "db.watchfunction"
        Called from within "DB::watchfunction()". If you want the debugger
        to call the function, you need to enable it by calling
        "enable_watchfunction()" somewhere within your plugin. It's a good
        idea to enable it as late as possible because it is being called
        very often. See the DB::Pluggable::BreakOnTestNumber source code for
        an example. The hook doesn't get any arguments.

    "db.cmd.b"
        Called when the "b" debugger command (used to set breakpoints) is
        invoked. See "run()" below for what the hook should return.

        The hook passes these named arguments:

        "cmd"
            This is the first argument passed to "DB::cmd_b()".

        "line"
            This is the second argument passed to "DB::cmd_b()". This is the
            most important argument as it contains the command line. See the
            DB::Pluggable::BreakOnTestNumber source code for an example.

        "dbline"
            This is the third argument passed to "DB::cmd_b()".

    "db.eval"
        The debugger's "eval()" function is overridden so we can hook into
        it. This is needed to define new debugger commands that take
        arguments. Each plugin that registered this hook will get a chance
        to inspect the command line, which is the last line in $DB::evalarg
        and act on it. The plugin can return a code reference which will be
        executed after the original "DB::eval()" function has finished.
        Using the code reference you can undo any temporary changes you
        might have introduced to make your command work. For example, read
        the source code of DB::Pluggable::Dumper.

INSTALLATION
    See perlmodinstall for information and options on installing Perl
    modules.

BUGS AND LIMITATIONS
    No bugs have been reported.

    Please report any bugs or feature requests through the web interface at
    <http://rt.cpan.org/Public/Dist/Display.html?Name=DB-Pluggable>.

AVAILABILITY
    The latest version of this module is available from the Comprehensive
    Perl Archive Network (CPAN). Visit <http://www.perl.com/CPAN/> to find a
    CPAN site near you, or see <http://search.cpan.org/dist/DB-Pluggable/>.

    The development version lives at
    <http://github.com/hanekomu/DB-Pluggable/>. Instead of sending patches,
    please fork this project using the standard git and github
    infrastructure.

AUTHOR
      Marcel Gruenauer <marcel@cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2008 by Marcel Gruenauer.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for Perl extension DB-Pluggable

1.112000  2011-07-19 20:30:08 Europe/Vienna
    - doc fix for the Data::Printer plugin (thanks garu)

1.111751  2011-06-24 23:54:05 Europe/Vienna
    - doc fix: don't mention nonexistent roles. :)

1.111750  2011-06-24 23:04:34 Europe/Vienna
    - complete rewrite to move from Hook::Modular to Brickyard - it's more
      lightweight and easier to use

1.101051  2010-04-15 18:01:53 Europe/Vienna
    - fixed the db.eval hook handler

1.101050  2010-04-15 00:15:37 Europe/Vienna
    - added db.eval hook so you can define custom commands that take arguments
    - removed unused variables and code

1.100851  2010-03-26 20:56:57 Europe/Vienna
    - DB::Pluggable::Plugin now returns a true value

1.100850  2010-03-26 20:50:12 Europe/Vienna
    - converted the distribution to Dist::Zilla-style

0.04  Thu Jul 25 15:27:43 CEST 2008 (Marcel Gruenauer <marcel@cpan.org>)
    - Changed dist style and Changes back to standard. People didn't like it -
      the nail that sticks out gets hammered down.
    - Added standard test files; this will also help with CPANTS
    - Added DB::Pluggable::Plugin base class for plugins

0.03  Thu, 29 May 2008 12:04:18 -0000 (Marcel Gruenauer <marcel@cpan.org>)
    - set the version to 0.03
    - fixed first year, is 2008
    - .shipit: fixed svk.tagpattern

0.02  Thu, 17 Apr 2008 21:49:33 -0000 (Marcel Gruenauer <marcel@cpan.org>)
    - hiding the DB package from the PAUSE indexer
    - set version to 0.02

0.01  Thu, 17 Apr 2008 19:45:15 -0000 (Marcel Gruenauer <marcel@cpan.org>)
    - original version


lib/DB/Pluggable/BreakOnTestNumber.pm
use 5.008;
use strict;
use warnings;

package DB::Pluggable::BreakOnTestNumber;
BEGIN {
  $DB::Pluggable::BreakOnTestNumber::VERSION = '1.101051';
}
# ABSTRACT: Debugger plugin to break on Test::Builder-based tests
use DB::Pluggable::Constants ':all';
use parent 'Hook::Modular::Plugin';

sub register {
    my ($self, $context) = @_;
    $context->register_hook(
        $self,
        'plugin.init'      => $self->can('plugin_init'),
        'db.cmd.b'         => $self->can('cmd_b'),
        'db.watchfunction' => $self->can('watchfunction'),
    );
}

sub plugin_init {
    @DB::testbreak = ();
}

sub cmd_b {
    my ($self, $context, $args) = @_;
    if ($args->{line} =~ /\s*#\s*(\d+(?:\s*,\s*\d+)*)$/) {
        my %seen;
        @DB::testbreak = grep { !$seen{$_}++ }
          sort { $a <=> $b } (split(/\s*,\s*/, $1), @DB::testbreak);

        # enable the watchfunction JIT so we don't waste time at the beginning
        $context->enable_watchfunction;
        return HANDLED;
    } else {
        return DECLINED;
    }
}

sub watchfunction {
    my $self = shift;
    if (@DB::testbreak && exists $INC{'Test/Builder.pm'}) {
        my $next = Test::Builder->new->current_test + 1;
        if ($next >= $DB::testbreak[0]) {
            shift @DB::testbreak
              while @DB::testbreak && $next >= $DB::testbreak[0];
            my $depth = 1;
            while (1) {
                my $package = (caller $depth)[0];
                last unless defined $package;
                last unless $package =~ /^(DB(::|$)|Test::)/;
                $depth++;
            }

            no warnings 'once';
            $DB::stack[ -$depth ] = 1;
        }
    }
    return;
}
1;


__END__
=pod

=for stopwords watchfunction

=for test_synopsis 1;
__END__

=head1 NAME

DB::Pluggable::BreakOnTestNumber - Debugger plugin to break on Test::Builder-based tests

=head1 VERSION

version 1.101051

=head1 SYNOPSIS

    $ cat ~/.perldb

    use DB::Pluggable;
    use YAML;

    $DB::PluginHandler = DB::Pluggable->new(config => Load <<EOYAML);
    global:
      log:
        level: error

    plugins:
      - module: BreakOnTestNumber
    EOYAML

    $DB::PluginHandler->run;

    $ perl -d foo.pl

    Loading DB routines from perl5db.pl version 1.28
    Editor support available.

    Enter h or `h h' for help, or `man perldebug' for more help.

    1..9
    ...
      DB<1> b #5
      DB<2> r

=head1 DESCRIPTION

This debugger plugin extends the debugger's C<b> command - used to set
breakpoints - with the ability to stop at a specific test number. Andy
Armstrong had the idea and wrote the original code, see
L<http://use.perl.org/~AndyArmstrong/journal/35792>.

=head1 METHODS

=head2 register

Registers the hooks.

=head2 plugin_init

Hook handler for the C<plugin.init> hook. Does some initializations,
surprisingly.

=head2 cmd_b

Hook handler for the C<db.cmd.b> hook. Checks whether the command is of the
form C<b #12> or C<b #12, 34, ...>. If so, it sets breakpoints to break as
soon as the code leading up to the given test is encountered. For example,
given C<b #12>, it breaks as soon as test 11 has been finished. If it has
handled the command, it returns the L<DB::Pluggable::Constants>
C<HANDLED>. If not, it returns C<DECLINED>.

If it handles the command, it enables the C<watchfunction()>.

=head2 watchfunction

Hook handler for the C<db.watchfunction> hook. Checks the current test number
from L<Test::Builder> and instructs the debugger to stop if an appropriate
test number has been reached.

=head1 INSTALLATION

See perlmodinstall for information and options on installing Perl modules.

=head1 BUGS AND LIMITATIONS

No bugs have been reported.

Please report any bugs or feature requests through the web interface at
L<http://rt.cpan.org/Public/Dist/Display.html?Name=DB-Pluggable>.

=head1 AVAILABILITY

The latest version of this module is available from the Comprehensive Perl
Archive Network (CPAN). Visit L<http://www.perl.com/CPAN/> to find a CPAN
site near you, or see
L<http://search.cpan.org/dist/DB-Pluggable/>.

The development version lives at
L<http://github.com/hanekomu/DB-Pluggable/>.
Instead of sending patches, please fork this project using the standard git
and github infrastructure.

=head1 AUTHOR

  Marcel Gruenauer <marcel@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2008 by Marcel Gruenauer.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut

